akka-actor-typed/src/main/scala/akka/actor/typed/internal/Restarter.scala

```diff
@@ -44,13 +45,46 @@ import scala.util.control.NonFatal
       supervisor.init(ctx)
     }
 
+  // find supervision that is superflous by having the same exception handled closer to the behavior
+  // and remove those instances. Needs to be called in the wrap method of all supervisor classes
+  def deduplicate[T, Thr <: Throwable: ClassTag](supervisor: Supervisor[T, Thr]): Supervisor[T, Thr] = {
+
+    // side effecting to avoid allocating a tuple per loop
+    var seenSupervised = Set.empty[Class[_]]
+
+    // can't be tailrec, but should be ok since hierarchies shouldn't be _that_ deep given that they
+    // are deduplicated for every wrap
+    def loop(behavior: Behavior[T]): Behavior[T] = {
+      behavior match {
+        case s: Supervisor[T, _] ⇒
+          val inner = loop(s.behavior)
+          if (seenSupervised.contains(s.throwableClass))
```

Comment:
we could discard subclasses also but makes it a bit more complicated and the main purpose is to remove the accidental wrapping of same

Reply:
Yeah, I thought about that, but wasn't sure it is worth because it isn't very realistic that you add more general and more general supervision over time.